### PR TITLE
reset evaluation after it started

### LINKED ISSUE #1991: Reset evaluation after it started
It is not currently possible to reset a running evaluation. If it turns out that the start of the evaluation period has been set too early, or if the evaluation has to be stopped for other reasons, a reset can currently only be done in the database.

The evaluation operation `revert_to_new` ("Revert to preparation" in the frontend), which resets the selected evaluation to state `new`, should be extended to other states. The operation should be allowed for all states from `prepared` to `reviewed` (all states except `new` and `published`) and should be called `reset_to_new` ("Reset evaluation" in the frontend).
When performing this operation, the user should be able to select a checkbox (default: checked) on the operation confirmation page to delete previously received answers.
If checked 
- all received answers (`TextAnswer` and `TextAnswerCounter`) for this evaluation will be deleted, and
- the `voters` list is cleared.

To prevent unintentional deletion of data, an additional confirmation modal should be displayed if the checkbox for the data to be deleted is set.

### PR DESCRIPTION
**Pull request by fekoch**
Closes #1991 
Replaces the `revert_to_new` evaluation operation with `reset_to_new`. It now allows resetting from any State except `NEW` and `PUBLISHED`. It also allows deleting all previously received answers.

Also, it makes `Evaluation.State` an `IntEnum`.

- [x] Only show confirmation modal when deleting previous answers
- [x] Finish Unit-Tests
- [x] Cleanup leftover TODO's

### CODE REVIEW

File: evap/staff/templates/staff_evaluation_operation.html
Comment (fekoch) [inline evap/staff/templates/staff_evaluation_operation.html:91]: @niklasmohrin As I understand it, you worked on the new `<confirmation-modal>`. Do you have an idea for a nice way to skip the modal or only show it when certain conditions are met?

Comment (niklasmohrin) [reply evap/staff/templates/staff_evaluation_operation.html:91]: The modal display is triggered here: https://github.com/e-valuation/EvaP/blob/4d998159bff118e5ac14529201cc8da9695e1e80/evap/static/ts/src/confirmation-modal.ts#L30

You could make this single line available as a method on `ConfirmationModal` and then use JavaScript to dynamically use the modal or `form.requestSubmit()` directly. Your JavaScript can then use an arbitrary condition for when to use which of the two.

Does this help?

File: evap/staff/templates/staff_evaluation_operation.html
Comment (niklasmohrin) [inline evap/staff/templates/staff_evaluation_operation.html:89]: The show-button should probably never be a submit button, because form elements defined here (in the users of the confirmation modal) will bind to a form outside of the modal, not within it. The show button should not submit any forms and therefore not be a submit type.

However, it may be practical to overwrite the submit row with buttons that belong to forms outside of the modal, so there it could make sense.

File: evap/staff/templates/staff_evaluation_operation.html
Comment (fekoch) [reply evap/staff/templates/staff_evaluation_operation.html]: I added the option to set `skipShowingModal`, a generic `()=>boolean`-function, on the modal. With this one can easily specify the logic next to the modal in question:

```html
<script>
    const modal = document.getElementById("confirmation-modal");
                                                                           
    function isDeleteAnswersUnchecked() {
        return !document.getElementById("delete-previous-answers").checked;
    }
                                                                           
    modal.skipShowingModal = isDeleteAnswersUnchecked;
</script>
```

What do you think?

Comment (niklasmohrin) [reply evap/staff/templates/staff_evaluation_operation.html]: After thinking about it again, maybe it would be better if "submit without modal" would not be the default. Instead, the extra button could just be ˋtype=buttonˋ and either trigger the modal, or do ˋform.requestSubmit() otherwise

**Comment by fekoch** [comment]
Two Unit Tests are missing. I do not have any idea on how to practically test for ui popups without selenium tests

File: evap/evaluation/migrations/0143_alter_evaluation_state.py
Comment (niklasmohrin) [inline evap/evaluation/migrations/0143_alter_evaluation_state.py:20]: Is this what came out of `makemigrations`? Usually it would emit `10` instead of `Evaluation.State["NEW"]` here, with a good reason: The imported `Evaluation` class always reflects the current state of the codebase, so if we change `Evaluation.State` in the future, this migration would also use the changed version. We want the migration to make changes according to what `Evaluation.State` meant at the of writing though. See for example the migration `0123_evaluation_state_fsm_int` where I changed evaluation states to be numbers instead of strings. There, because the migration should work regardless of the current definition of `Evaluation`, all constants are copied in the migration

**Review by niklasmohrin — COMMENTED** [review]
I started reviewing your changes and I find it a bit tricky to separate the "revert -> reset" and "make State an IntEnum" changes. Do you think you could split these changes into two PRs? If not, we will have to see how we manage :D 

The extra PR would also be a good place to discuss the motivation for the move to `IntEnum` - it sounds like a good idea, but it's hard to see the benefits when they are mixed in with the other changes here

**Comment by fekoch** [comment]
@fekoch some TODO's:

- [ ] type annotation in test helper function
- [ ] remove empty newline

File: evap/evaluation/models.py
Comment (richardebeling) [inline evap/evaluation/models.py:621]: I'd say this is probably fine without the comment, and generally, I'll vote to remove any comments that don't add value.

File: evap/evaluation/models.py
Comment (richardebeling) [inline evap/evaluation/models.py:732]: Why `| None`?

File: evap/evaluation/models.py
Comment (richardebeling) [inline evap/evaluation/models.py:733]: Maybe also remove? The history/background/issue/PR are available in the introducing commits.

File: evap/staff/templates/staff_evaluation_operation.html
Comment (richardebeling) [inline evap/staff/templates/staff_evaluation_operation.html:86]: Should probably not be in the `question` slow. Can you add a short comment that the modal always expects this element to exist, but we don't need it on this page?

(I'd also be fine with modifying the confirmation modal to have nicely if an element for this slot does not exist)

File: evap/evaluation/models.py
Comment (fekoch) [reply evap/evaluation/models.py:732]: So that one doesn't have to specify `delete_previous_answers=False` for every `.reset_to_new()`. But as every other current usage is in the tests, it will be clearer when this is removed

File: evap/evaluation/models.py
Comment (richardebeling) [inline evap/evaluation/models.py:731]:
```suggestion
    def reset_to_new(self, *, delete_previous_answers: bool):
```

**Review by richardebeling — APPROVED** [review]

File: evap/staff/templates/staff_evaluation_operation.html
Comment (janno42) [inline evap/staff/templates/staff_evaluation_operation.html]:
```suggestion
                <button id="form-confirmation-btn" type="button"
                        class="btn btn-danger">{% translate 'Confirm' %}</button>
```

File: evap/staff/templates/staff_evaluation_operation.html
Comment (janno42) [inline evap/staff/templates/staff_evaluation_operation.html:83]:
```suggestion
                            This will delete all previously received answers. Voters will have to resubmit the
                            evaluation.
```

File: evap/staff/views.py
Comment (janno42) [inline evap/staff/views.py:300]:
```suggestion
            "{} evaluation cannot be reset, because it is already in preparation or published. It was removed from the selection",
            "{} evaluations cannot be reset, because they were already in preparation or published. They were removed from the selection.",
```

**Review by janno42 — CHANGES_REQUESTED** [review]

**Review by janno42 — CHANGES_REQUESTED** [review]
- Forgot to specify before: It should not be possible to reset single results. They should be removed from the selection and the operation should not be `applicable_to` them.
- Submitting other operations (e.g., start evaluation, publish, unpublish) is not possible anymore. The submit button does not send a request in these cases (`document.getElementById("delete-previous-answers")` fails).

File: evap/staff/views.py
Comment (Kakadus) [inline evap/staff/views.py:317]: as noted, it is a BIT sus that we pass `bool | None` to `reset_to_new`, but I think it's okayish here

**Review by Kakadus — APPROVED** [review]

File: evap/staff/templates/staff_evaluation_operation.html
Comment (fekoch) [reply evap/staff/templates/staff_evaluation_operation.html]: image

Ist das wirklich der Button den du meinst?

Comment (fekoch) [reply evap/staff/templates/staff_evaluation_operation.html]: Oder meinst du nicht eher den Bestätigungsbutton?

Comment (fekoch) [reply evap/staff/templates/staff_evaluation_operation.html]: image

File: evap/staff/views.py
Comment (richardebeling) [reply evap/staff/views.py:317]: I'd prefer the narrower contract on the function, i.e. annotating the function as only taking in a `bool` and then explicitly casting the possibly None value to bool on the caller site. What do you think about that?

Comment (fekoch) [reply evap/staff/views.py:317]: Yeah, we can just do it narrower. About casting on the caller site:

The only caller site is currently the `ResetToNewOperation`. We could do it like:

```python
evaluation.reset_to_new(delete_previous_answers=bool(delete_previous_answers))
```

But as `delete_previous_answers` is not currently typed inside `ResetToNewOperation.apply()`, nothing complains if we still just do:
```python
evaluation.reset_to_new(delete_previous_answers=delete_previous_answers)
```

Comment (fekoch) [reply evap/staff/views.py]: @richardebeling @Kakadus I completely removed the inheritance and instead use types to validate that the correct method is called.

I am also thinking of adding a test like 
```python
class TestWrongOperationArgs(TestCase):
    def test_wrong_args_throw_typeerror(self):
        evaluation = baker.make(
            Evaluation,
            state=Evaluation.State.PREPARED
        )
        _, operation = evap.staff.views.target_state_and_operation_from_str(str(Evaluation.State.NEW))

        args = {"email_template": EmailTemplate()}

        self.assertRaisesRegex(
            TypeError,
            "got an unexpected keyword argument",
            operation.apply,
            Mock(), [evaluation], **args
        )
```

Do you think this makes sense? Is this not needed? Or should I cover more/all of the operations?

**Review by janno42 — CHANGES_REQUESTED** [review]

File: evap/staff/templates/staff_evaluation_operation.html
Comment (niklasmohrin) [inline evap/staff/templates/staff_evaluation_operation.html:98]: I think we can clean this up a bit, I imagine the following structure:

```html
<form id="the-form">
    {% if something %}
        <input type="checkbox" name="delete-previous-answers" checked />
    {% endif %}

    <confirmation-modal type="submit" data-allow-answer-deletion ...>
        ...
    </confirmation-modal>

    <button type="submit" ...></button>
</form>

<script>
    const form = document.getElementById("the-form");
    form.addEventListener("submit", event => {
        const data = new FormData(form);
        if (data.get("delete-previous-answers") && !event.submitter?.hasAttribute("data-allow-answer-deletion")) {
            event.preventDefault();
            form.querySelector("confirmation-modal[data-allow-answer-deletion]").dialog.showModal();
        }
    });
</script>
```

This way, we avoid more `id`s and don't need to include different buttons depending on what value `show_delete_answers_checkbox` has. What do you think?

File: evap/staff/templates/staff_evaluation_operation.html
Comment (niklasmohrin) [inline evap/staff/templates/staff_evaluation_operation.html:65]: If possible, I favor nesting inputs and labels instead of giving out `id`s. By putting the `input` inside the `label`, they also become associated without having to use `id` and `for`. (see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label)

File: evap/staff/templates/staff_evaluation_operation.html
Comment (niklasmohrin) [inline evap/staff/templates/staff_evaluation_operation.html:74]: Since this element is already nested in the `evaluation-operation-form`, the `form=` attribute is not needed :)

**Review by niklasmohrin — CHANGES_REQUESTED** [review]
Some notes:

File: evap/evaluation/models.py
Comment (fekoch) [reply evap/evaluation/models.py:620]: Seems good, I used it the other way around because the logic was simple enough. But it will be probably more efficient the other way around.

File: evap/staff/templates/staff_evaluation_operation.html
Comment (fekoch) [reply evap/staff/templates/staff_evaluation_operation.html:9]: What do you think about removing the slash altogether? For the `<input>` and `<col>` it is not needed

Comment (richardebeling) [reply evap/staff/templates/staff_evaluation_operation.html:9]: I think for general parsing as a human, it's a bit nicer to have the slashes there, indicating that the tag immediately closes, so I'd leave them.

File: evap/staff/templates/staff_evaluation_operation.html
Comment (fekoch) [reply evap/staff/templates/staff_evaluation_operation.html:98]: @niklasmohrin I changed it to this. We need to put some kind of marker on the button inside the modal as this element will be used the `.submitter` value.

```html
<form id="evaluation-operation-form">
	<confirmation-modal type="submit" data-allow-answer-deletion>
	...
	    <button slot="submit-group" class="btn ms-2 btn-danger" data-event-type="confirm" data-confirm-answer-deletion>
		{% translate 'Reset Evaluation' %}
	    </button>

	    <span slot="show-button"></span>
	</confirmation-modal>

	<button type="submit" class="btn btn-primary">{% translate 'Confirm' %}</button>
	<script>
	    document.getElementById("evaluation-operation-form").addEventListener("submit", submitEvent => {
		const data = new FormData(submitEvent.target);
		if (data.get("delete-previous-answers") === "on" && !submitEvent.submitter?.hasAttribute("data-confirm-answer-deletion")) {
		    submitEvent.preventDefault()
		    submitEvent.target.querySelector("confirmation-modal[data-allow-answer-deletion]").dialog.showModal();
		}
	    })
	</script>
</form>
```

But I am not sure if I like it, because now the modal and javascript is used on every evaluation operation.

Comment (niklasmohrin) [reply evap/staff/templates/staff_evaluation_operation.html]: Ough I forgot about this: https://github.com/e-valuation/EvaP/blob/a9e417180b6a8e380bbcb9885b301f13e9524bc0/evap/static/ts/src/confirmation-modal.ts#L53-L55
(https://github.com/WICG/webcomponents/issues/814)

I wrote the above thinking that the modal itself would be the submitter :/ 

> But I am not sure if I like it, because now the modal and javascript is used on every evaluation operation.

I am not sure I understand this point - the javascript runs on every submit, but that is not noticable if the condition in the handler is `false`. I prefer handling things in the `submit` events over `click` events because it marks the form as the central actor of the process. The modal is only shown if the delete-answers checkbox is checked, so the modal should only get involved when we really want to ask for additional confirmation - or does it appear more often than needed now? Or am I not getting something else here?

---

I think your solution with the explicit `submit-group` is the only correct way if we want a solution where all _allowed_ submitters are marked with a data attribute. We could also do it the other way around and mark the other submit button as (for example) `data-prompt-confirmation-for-delete-answers` - and we leave the modal just as is. This approach is slightly less nice beacuse the default submission is "unsafe". However, there are only two possible submission sources for the operation form, so this should be fine.

What do you think?

Comment (fekoch) [reply evap/staff/templates/staff_evaluation_operation.html]: re 2: Yeah, I would also go for the current way. It just seems more explicit & readable that way.

re 1: My point was that this way we always send and execute additional javascript to the client, when it is only used for one of the operations. But this was only a minor thought.

**Review by janno42 — APPROVED** [review]
✔️ Meets requirements
✔️ UI functionality checked

**Review by richardebeling — APPROVED** [review]
thanks :)

**Review by Kakadus — APPROVED** [review]

File: evap/evaluation/models.py
Comment (niklasmohrin) [inline evap/evaluation/models.py:630]: Any reason to prefer this over the following?
```suggestion
        return Evaluation.State.NEW < self.state < Evaluation.State.PUBLISHED and not self.is_single_result
```

File: evap/evaluation/models.py
Comment (niklasmohrin) [inline evap/evaluation/models.py:730]: Does the following also work?
```suggestion
                answer_class._default_manager.filter(contribution__evaluation=self).delete()
```

File: evap/staff/views.py
Comment (niklasmohrin) [inline evap/staff/views.py:584]: Couldn't we do `delete_previous_answers = request.POST.get("delete-previous-answers") == "on"` from the start? I think this would also fix the `bool | None` "issue", although I don't care so much about that

File: evap/staff/templates/staff_evaluation_operation.html
Comment (niklasmohrin) [inline evap/staff/templates/staff_evaluation_operation.html:87]: Good catch, we should change the modal so that this is not needed, you or me can fix it in a follow up :)

File: evap/staff/templates/staff_evaluation_operation.html
Comment (niklasmohrin) [inline evap/staff/templates/staff_evaluation_operation.html:95]: With the following, we can get rid of the `data-allow-answer-deletion` on the modal:
```suggestion
                            submitEvent.target.querySelector("confirmation-modal:has([data-confirm-answer-deletion])").dialog.showModal();
```
You can decide which version you prefer, I think both are fine

**Review by niklasmohrin — APPROVED** [review]
Nice job, I have some final remarks below.

Since all reviews are green now, but you still may want to change something, just write a quick comment once you are okay with merging

File: evap/staff/views.py
Comment (fekoch) [reply evap/staff/views.py:586]: We could do it like that, but then we would need to change the `apply` methods of all other operations to expect `False` instead of `None`. This is doable, but as all the other parameters use `None` I do not think it will contribute to readability.

```python
    @staticmethod
    def apply(
        request,
        evaluations,
        email_template=None,
        email_template_contributor=None,
        email_template_participant=None,
        delete_previous_answers=None,
    ):
        assert email_template_contributor is None
        assert email_template_participant is None
        assert delete_previous_answers is None
        
        # ....
```

 So I would argue for keeping it as it is now.

File: evap/staff/templates/staff_evaluation_operation.html
Comment (fekoch) [reply evap/staff/templates/staff_evaluation_operation.html:95]: this seems very neat

File: evap/staff/views.py
Comment (richardebeling) [reply evap/staff/views.py:586]: I think the other `apply` methods requiring `None` is fine (not type-annotated in the current version, and subject to change anyway?), but we can still use `delete_previous_answers = request.POST.get("delete-previous-answers") == "on"` here.

**Review by richardebeling — APPROVED** [review]

File: evap/staff/views.py
Comment (niklasmohrin) [reply evap/staff/views.py:586]: either way is fine - if we are divided on this, we can also leave it as is for now

**Review by niklasmohrin — APPROVED** [review]
Very cool, thanks!

**Pull request merged**
